packages/wrangler/src/type-generation.ts

```diff
@@ -1,16 +1,115 @@
-import * as fs from "fs";
+import * as fs from "node:fs";
 import { findUpSync } from "find-up";
+import { findWranglerToml, readConfig } from "./config";
 import { getEntry } from "./deployment-bundle/entry";
+import { getVarsForDev } from "./dev/dev-vars";
 import { UserError } from "./errors";
 import { logger } from "./logger";
+import { printWranglerBanner } from "./update-check";
+import { CommandLineArgsError } from "./index";
 import type { Config } from "./config";
 import type { CfScriptFormat } from "./deployment-bundle/worker";
+import type {
+	CommonYargsArgv,
+	StrictYargsOptionsToInterface,
+} from "./yargs-types";
 
-// Currently includes bindings & rules for declaring modules
+export function typesOptions(yargs: CommonYargsArgv) {
+	return yargs
+		.positional("path", {
+			describe: "The path to the declaration file to generate",
+			type: "string",
+			default: "worker-configuration.d.ts",
+			demandOption: false,
+		})
+		.option("env-interface", {
+			type: "string",
+			default: "Env",
+			describe: "The name of the generated environment interface",
+			requiresArg: true,
+		});
+}
 
-export async function generateTypes(
-	configToDTS: Partial<Config>,
-	config: Config
+export async function typesHandler(
+	args: StrictYargsOptionsToInterface<typeof typesOptions>
+) {
+	const { envInterface, path: outputPath } = args;
+
+	const validInterfaceRegex = /^[a-zA-Z][a-zA-Z0-9_]*$/;
+
+	if (!validInterfaceRegex.test(envInterface)) {
+		throw new CommandLineArgsError(
+			`The provided env-interface value ("${envInterface}") does not satisfy the validation regex: ${validInterfaceRegex}`
+		);
+	}
+
+	if (!outputPath.endsWith(".d.ts")) {
+		throw new CommandLineArgsError(
+			`The provided path value ("${outputPath}") does not point to a declaration file (please use the 'd.ts' extension)`
+		);
+	}
+
+	await printWranglerBanner();
+
+	const configPath =
+		args.config ?? findWranglerToml(process.cwd(), args.experimentalJsonConfig);
+	if (
+		!configPath ||
+		!fs.existsSync(configPath) ||
+		fs.statSync(configPath).isDirectory()
+	) {
+		logger.warn(
+			`No config file detected${
+				args.config ? ` (at ${args.config})` : ""
+			}, aborting`
+		);
+		return;
+	}
+
+	const config = readConfig(args.config, args);
+
+	const configBindings: Partial<Config> = {
+		kv_namespaces: config.kv_namespaces ?? [],
+		vars: { ...config.vars },
+		wasm_modules: config.wasm_modules,
+		text_blobs: {
+			...config.text_blobs,
```

Comment:
Out of interest, do you know why we are cloning this object?

Reply:
now that you mention it, we have:
```ts
vars: { ...config.vars },
```
as well as:
```ts
kv_namespaces: config.kv_namespaces
```
and
```ts
durable_objects: config.durable_objects
```

etc...

a bit inconsistent 😕 (and also unnecessary I feel? 🤷)